### PR TITLE
[Base] add electrical conductivity

### PR DESCRIPTION
**Pull request by donovaly**
- this is necessary to perform electrodynamics simulations. Since the typical values are in the range of 1e6 S/m, we need the new units kS and MS

- since this is for simulations, also add electrical conductivity to the FEM material cards

### CODE REVIEW

File: src/Base/QuantityLexer.c [hunk 1]
```diff
@@ -56,7 +56,7 @@
 #if defined (__STDC_VERSION__) && __STDC_VERSION__ >= 199901L
 
 /* C99 says to define __STDC_LIMIT_MACROS before including stdint.h,
- * if you want the limit (max/min) macros for int types.
+ * if you want the limit (max/min) macros for int types.
```
Comment (luzpaz) [inline src/Base/QuantityLexer.c:59]: no need for trailing whitespace

File: src/Base/QuantityLexer.c [hunk 2]
```diff
@@ -73,7 +73,7 @@ typedef uint32_t flex_uint32_t;
 typedef signed char flex_int8_t;
 typedef short int flex_int16_t;
 typedef int flex_int32_t;
-typedef unsigned char flex_uint8_t;
+typedef unsigned char flex_uint8_t;
```
Comment (luzpaz) [inline src/Base/QuantityLexer.c:76]: remove trailing whitespace

File: src/Base/QuantityLexer.c [hunk 3]
```diff
@@ -184,10 +184,10 @@ extern FILE *yyin, *yyout;
 #define EOB_ACT_CONTINUE_SCAN 0
 #define EOB_ACT_END_OF_FILE 1
 #define EOB_ACT_LAST_MATCH 2
-
+
```
Comment (luzpaz) [inline src/Base/QuantityLexer.c:187]: remove superflous whitespace (also below this)

File: src/Base/QuantityLexer.c [hunk 4]
```diff
@@ -681,7 +683,7 @@ char *yytext;
  */
 #include <unistd.h>
 #endif
-
+
```
Comment (luzpaz) [inline src/Base/QuantityLexer.c:686]: remove whitespace

File: src/Base/QuantityLexer.c [hunk 5]
```diff
@@ -857,7 +859,7 @@ YY_DECL
 	yy_state_type yy_current_state;
 	char *yy_cp, *yy_bp;
 	int yy_act;
-
+
```
Comment (luzpaz) [inline src/Base/QuantityLexer.c:862]: remove pls

File: src/Base/QuantityLexer.c [hunk 6]
```diff
@@ -1992,7 +2004,7 @@ static int yy_get_next_buffer (void)
 {
 	yy_state_type yy_current_state;
 	char *yy_cp;
-
+
```
Comment (luzpaz) [inline src/Base/QuantityLexer.c:2007]: remove

File: src/Base/QuantityLexer.c (hunk 1 above)
Comment (donovaly) [reply src/Base/QuantityLexer.c:59]: This file is automatically generated by the program flex. So we can ignore these whitespace issues.

**Review by luzpaz — CHANGES_REQUESTED** [review]
there are other added superfluous whitespace. Please remove. Thank

**Comment by donovaly** [comment]
> there are other added superfluous whitespace. Please remove. Thank

As I wrote, QuantityLexer.c is automatically generated by flex. So we can ignore its formatting.

**Comment by berndhahnebach** [comment]
ok from my side.

**Comment by donovaly** [comment]
> ok from my side but I have no idea what this QuantityLexer.c is about. Thus I requested Werner to review.

Thanks for the review. @wwmayer  OK for you as well?

QuantityLexer.c is generated by flex (a free implementation of https://en.wikipedia.org/wiki/Lex_(software) ). The source code for the generation is the file QuantityParser.l  and this is our lexer (parser) for quantities.

File: src/App/FreeCADInit.py [hunk 7]
```diff
@@ -807,6 +809,7 @@ def report(self,msg,func,*args,**kargs):
 App.Units.ElectricalInductance  = App.Units.Unit(2,1,-2,-2)
 App.Units.ElectricalConductance = App.Units.Unit(-2,-1,3,2)
 App.Units.ElectricalResistance  = App.Units.Unit(2,1,-3,-2)
+App.Units.ElectricalConductivity = App.Units.Unit(-3,-1,3,2)
```
Comment (wwmayer) [inline src/App/FreeCADInit.py:812]: According to https://de.wikipedia.org/wiki/Siemens_(Einheit) this unit is wrong. It must be (-2,-1,3,2)

File: src/Base/Unit.h [hunk 8]
```diff
@@ -121,6 +121,7 @@ class BaseExport Unit
     static Unit ElectricalInductance;
     static Unit ElectricalConductance;
     static Unit ElectricalResistance;
+    static Unit ElectricalConductivity;
```
Comment (wwmayer) [inline src/Base/Unit.h:124]: Two lines above ElectricalConductance is already declared which I assume is the same thing as ElectricalConductivity

File: src/Base/UnitsSchemaInternal.cpp [hunk 9]
```diff
@@ -360,6 +360,24 @@ QString UnitsSchemaInternal::schemaTranslate(const Quantity &quant, double &fact
             factor = 1e12;
         }
     }
+    else if (unit == Unit::ElectricalConductivity) {
```
Comment (wwmayer) [inline src/Base/UnitsSchemaInternal.cpp:363]: These factors are all wrong. When you have S/m then you must express this with the internal units where length is measured in mm.
1 S/m = A² * s³ / (kg * m³) = A² * s³ / (kg * (1000 mm)³) = 10⁻⁹ A² * s³ / (kg * mm³)

So the factors are:
mS/m -> 1e-12
S/m -> 1e-9
kS/m -> 1e-6
MS/m -> 1e-3

**Comment by wwmayer** [comment]
Fixed & Merged.

File: src/Base/UnitsSchemaInternal.cpp (hunk 9 above)
Comment (donovaly) [reply src/Base/UnitsSchemaInternal.cpp:363]: > These factors are all wrong. 
> 1 S/m = A² * s³ / (kg * m³)

Thanks! How embarrassing for me. I only cared about the denominator despite Siemens itself is a combined unit.

**Comment by berndhahnebach** [comment]
damn I missed that too ....

**Pull request closed without merging**